app/modules/somfy.c

```diff
@@ -0,0 +1,246 @@
+#define SOMFY_PROG  0x8
+
+#define DIRECT_WRITE_LOW(pin)    (GPIO_OUTPUT_SET(GPIO_ID_PIN(pin_num[pin]), 0))
+#define DIRECT_WRITE_HIGH(pin)   (GPIO_OUTPUT_SET(GPIO_ID_PIN(pin_num[pin]), 1))
+
+static const os_param_t TIMER_OWNER = 0x736f6d66; // "somf"
+static task_handle_t done_taskid;
+
+static uint8_t    pin;
+static uint8_t    frame[7];
+static uint8_t    sync;
+static uint8_t    repeat;
+
+//static uint32_t   delay[10] = {9415, 89565, 4*SYMBOL, 4*SYMBOL, 4*SYMBOL, 4550, SYMBOL, SYMBOL, SYMBOL, 30415}; // in us
+static const __attribute__((section(".data.delay"))) uint32_t delay[10] = {US_TO_RTC_TIMER_TICKS(9415), US_TO_RTC_TIMER_TICKS(89565), US_TO_RTC_TIMER_TICKS(4*SYMBOL), US_TO_RTC_TIMER_TICKS(4*SYMBOL), US_TO_RTC_TIMER_TICKS(4*SYMBOL), US_TO_RTC_TIMER_TICKS(4550), US_TO_RTC_TIMER_TICKS(SYMBOL), US_TO_RTC_TIMER_TICKS(SYMBOL), US_TO_RTC_TIMER_TICKS(SYMBOL), US_TO_RTC_TIMER_TICKS(30415)}; // in ticks (no need to recalulate)
+
+//static uint32_t*  pdelay; //pointer to store delay array in ram
+
+static uint8_t    repeatindex;
+static uint8_t    signalindex;
+static uint8_t    subindex;
+static uint8_t    bitcondition;
+
+int lua_done_ref; // callback when transmission is done
+
+void buildFrame(uint8_t *frame, uint64_t remote, uint8_t button, uint16_t code) {
+    // NODE_DBG("remote: %x\n", remote);
+    // NODE_DBG("button: %x\n", button);
+    // NODE_DBG("rolling code: %x\n", code);
+    frame[0] = 0xA7; // Encryption key. Doesn't matter much
+    frame[1] = button << 4; // Which button did you press? The 4 LSB will be the checksum
+    frame[2] = code >> 8; // Rolling code (big endian)
+    frame[3] = code;         // Rolling code
+    frame[4] = remote >> 16; // Remote address
+    frame[5] = remote >> 8; // Remote address
+    frame[6] = remote;         // Remote address
+    // frame[7] = 0x80;
+    // frame[8] = 0x0;
+    // frame[9] = 0x0;
+
+    // NODE_DBG("Frame:\t\t\t%02x %02x %02x %02x %02x %02x %02x\n", frame[0], frame[1], frame[2], frame[3], frame[4], frame[5], frame[6]);
+    // Checksum calculation: a XOR of all the nibbles
+    uint8_t checksum = 0;
+    for(uint8_t i = 0; i < 7; i++) {
+        checksum = checksum ^ frame[i] ^ (frame[i] >> 4);
+    }
+    checksum &= 0b1111; // We keep the last 4 bits only
+
+    //Checksum integration
+    frame[1] |= checksum; //    If a XOR of all the nibbles is equal to 0, the blinds will consider the checksum ok.
+    // NODE_DBG("With checksum:\t%02x %02x %02x %02x %02x %02x %02x\n", frame[0], frame[1], frame[2], frame[3], frame[4], frame[5], frame[6]);
+
+    // Obfuscation: a XOR of all the uint8_ts
+    for(uint8_t i = 1; i < 7; i++) {
+        frame[i] ^= frame[i-1];
+    }
+    // NODE_DBG("Obfuscated:\t\t%02x %02x %02x %02x %02x %02x %02x\n", frame[0], frame[1], frame[2], frame[3], frame[4], frame[5], frame[6]);
+}
+
+static void somfy_transmissionDone (task_param_t arg)
+{
+    lua_State *L = lua_getstate();
+    lua_rawgeti (L, LUA_REGISTRYINDEX, lua_done_ref);
```

Comment:
Nope, it will indeed not. What you want to do is `luaL_unref()` as soon as you've done the `lua_rawgeti()`, or well, anywhere before you `lua_call()` at least. So something like
```lua
    lua_State *L = lua_getstate();
    lua_rawgeti (L, LUA_REGISTRYINDEX, lua_done_ref);
    luaL_unref (L, LUA_REGISTRYINDEX, lua_done_ref);
    lua_done_ref = LUA_NOREF;
    lua_call (L, 0, 0);
```
This way, the callback function is able to issue another somfy command successfully (since lua_done_ref is LUA_NOREF at that point), and the callback itself can become available for gc as soon as the `lua_call()` returns. While the callback is on the stack (after `lua_rawgeti()`), the stack itself holds a reference on it and thus ensures it doesn't get gc'd.